### PR TITLE
[DSCP-177] Fix lost heartbeat issue on the networking start

### PR DESCRIPTION
**Pull request by volhovm**
Pls don't merge yet

### CODE REVIEW

File: code/network/lib/Loot/Network/ZMQ/Adapter.hs [hunk 1]
```diff
@@ -6,65 +6,99 @@
...
+data SocketAdapter = SocketAdapter (TVar SocketState) (A.Async ())
 
 -- | Produces stm action corresponding to the action "lock until it is
 -- possible to read from the socket". It doesn't guarantee that there
 -- is actually any data to read from it.
-threadWaitReadSTMLong :: Fd -> IO (STM (), STM Bool, IO ())
+threadWaitReadSTMLong :: Fd -> IO (TVar SocketState, A.Async ())
 threadWaitReadSTMLong fd = do
     -- Bool value inside indicates who should take action now. If it's
     -- false, it's updater thread, if it's true it's main wait thread.
```
Comment (Heimdell) [inline code/network/lib/Loot/Network/ZMQ/Adapter.hs]: The comment is osbolete.

File: code/network/lib/Loot/Network/ZMQ/Adapter.hs [hunk 2]
```diff
@@ -6,65 +6,99 @@
...
-        atomically $ do
-            v <- readTMVar m
-            if v == False
-            then void (takeTMVar m)
-            else retry
+        -- Section 1
+        atomically $ readTVar m >>= \case
+            SSWaiting -> error "threadWaitReadSTMLong: can't happen"
+            SSCanRead -> retry
+            _         -> writeTVar m SSWaiting
```
Comment (Heimdell) [inline code/network/lib/Loot/Network/ZMQ/Adapter.hs]: Can we have `state --condition-> outcome` scheme for this state machine?

Also, you write `SSWaiting` into var, then on the next cycle if it is there you `error`. This construction seems brittle.

File: code/network/lib/Loot/Network/ZMQ/Client.hs [hunk 3]
```diff
@@ -363,33 +377,31 @@ runBroker = do
...
+    liftIO $ A.withAsync (heartbeatWorker cEnv) $ const $ forever $ do
+        results <- atomically $ do
+            cMap <- readTVar ztClients
+            let readReq =
+                    fmap CBRequest <$>
+                    TQ.tryReadTQueue (unCliRequestQueue ztCliRequestQueue)
+            let readClient :: [STM (Maybe CliBrokerStmRes)]
+                readClient =
+                    map (\(cliId,biq) ->
+                            fmap (\(nodeIdM,content) -> CBClient cliId nodeIdM content) <$>
```
Comment (Heimdell) [inline code/network/lib/Loot/Network/ZMQ/Client.hs]: can be `fmap (uncurry CBClient)`

File: code/network/lib/Loot/Network/ZMQ/Adapter.hs (hunk 2 above)
Comment (volhovm) [reply code/network/lib/Loot/Network/ZMQ/Adapter.hs]: I've added the explanation to the `SocketState` haddock.

File: code/network/lib/Loot/Network/ZMQ/Server.hs [hunk 4]
```diff
@@ -109,7 +115,9 @@ createNetServEnv (ZTGlobalEnv ctx ztLogging) ztOurNodeId = liftIO $ do
 -- | Terminates server environment.
 termNetServEnv :: MonadIO m => ZTNetServEnv -> m ()
 termNetServEnv ZTNetServEnv{..} = liftIO $ do
+    adapterRelease ztServFrontAdapter
     Z.close ztServFront
+    adapterRelease ztServPubAdapter
```
Comment (Martoon-00) [inline code/network/lib/Loot/Network/ZMQ/Server.hs]: It can receive async exception even under `finally` since `A.cancel` falls into interruptible sleep

**Pull request merged**